File: java/147.insertion-sort-list.java
/*
 * @lc app=leetcode id=147 lang=java
 *
 * [147] Insertion Sort List
 */
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode insertionSortList(ListNode head) {
        if (head == null || head.next == null) return head;
        ListNode be4Head = new ListNode(0);
        ListNode prev = be4Head, curr = head, next = null;
        while (curr != null) {
            next = curr.next;
            while (prev.next != null && 
                prev.next.val < curr.val) prev = prev.next;
            curr.next = prev.next;
            prev.next = curr;
            prev = be4Head;
            curr = next;
        }
        return be4Head.next;
    }

}

File: go/147.insertion-sort-list.go
/*
 * @lc app=leetcode id=147 lang=golang
 *
 * [147] Insertion Sort List
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func insertionSortList(head *ListNode) *ListNode {
	if head == nil {
		return head
	}
	ahead := &ListNode{0, head}
	head, head.Next = head.Next, nil
OUTER:
	for head != nil {
		prev, cur := ahead, ahead.Next
		for cur != nil {
			if head.Val < cur.Val {
				prev.Next = head
				head, head.Next = head.Next, cur
				continue OUTER
			}
			prev, cur = cur, cur.Next
		}
		prev.Next = head
		head, head.Next = head.Next, nil
	}
	return ahead.Next
}
// @lc code=end

